File: apple/App/SharedProfileSyncCoordinator.swift
import Domain
import Foundation
import Store
import SwiftData
import Sync

/// Doc 16, phase E — historique partagé : chaque partie terminée avec un ami lié lui est déposée,
/// **complète** et chiffrée, dans sa boîte aux lettres (`MatchMailboxTransport`) ; les parties que
/// d'autres m'ont déposées sont enregistrées ici comme si je les avais jouées. Remplace les
/// résumés du doc 14. Déclenché au lancement et au retour au premier plan (`QuiMeneApp`), dès
/// qu'une partie se termine (`ResultsView`, dépôt immédiat), et à l'ouverture de l'Historique (avec
/// « tirer pour actualiser ») pour relever sa boîte sans relancer l'app.
@MainActor
@Observable
final class SharedProfileSyncCoordinator {
  static let shared = SharedProfileSyncCoordinator()

  /// Change à chaque partie reçue : l'Historique affiché se recharge.
  private(set) var receivedToken = UUID()

  @ObservationIgnored private let transport = MatchMailboxTransport()
  @ObservationIgnored private var isSyncing = false
  @ObservationIgnored private var needsAnotherPass = false

  private init() {}

  /// Une demande pendant un passage en cours n'est pas perdue : un passage de plus suit.
  func sync(context: ModelContext) async {
    guard !isSyncing else {
      needsAnotherPass = true
      return
    }
    isSyncing = true
    defer { isSyncing = false }
    repeat {
      needsAnotherPass = false
      await deposit(context: context)
      await collect(context: context)
    } while needsAnotherPass
  }

  /// Une partie terminée avec au moins un ami lié (pas moi) : déposée chez chacun, une fois.
  /// Jamais retentée indéfiniment pour une partie qui n'en a plus (fiche déliée entre-temps).
  private func deposit(context: ModelContext) async {
    let repository = MatchRepository(context: context)
    guard let pending = try? repository.matchesPendingSharedProfileSync(), !pending.isEmpty
    else { return }
    let myID = (try? PlayerRepository(context: context).myOwnSharedPlayer())?.sharedProfileID

    for match in pending {
      let recipients = Set(match.participants.compactMap { $0.player?.sharedProfileID })
        .subtracting([myID].compactMap { $0 })
      guard !match.isImportedSummary, !recipients.isEmpty,
        let events = try? repository.currentLog(for: match), !events.isEmpty
      else {
        try? repository.markSharedProfileSyncComplete(match)
        continue
      }
      let package = SharedMatchPackage(
        matchID: match.id,
        participants: match.participants.map(Self.packaged),
        events: events)
      let items = recipients.compactMap { recipient in
        (try? MailboxCrypto.seal(package, for: recipient)).map {
          MailboxItem(
            mailboxKey: MailboxCrypto.lookupKey(for: recipient), matchID: match.id, ciphertext: $0)
        }
      }
      do {
        try await transport.deposit(items)
        try? repository.markSharedProfileSyncComplete(match)
      } catch {
        // Échec silencieux : `pendingSharedProfileSync` reste `true`, nouvelle tentative au
        // prochain retour au premier plan.
      }
    }
  }

  /// Ma boîte : chaque partie reçue est enregistrée (sans effet si je l'ai déjà, jouée ici ou
  /// suivie dans une session), puis retirée. Un dépôt illisible est retiré aussi, pour ne pas
  /// être relu indéfiniment.
  private func collect(context: ModelContext) async {
    guard let myID = (try? PlayerRepository(context: context).myOwnSharedPlayer())?.sharedProfileID
    else { return }
    let mailboxKey = MailboxCrypto.lookupKey(for: myID)
    guard let items = try? await transport.fetch(mailboxKey: mailboxKey) else { return }
    let repository = MatchRepository(context: context)
    var received = false
    for item in items {
      if let package = MailboxCrypto.open(item.ciphertext, for: myID) {
        guard (try? repository.importSharedMatch(package, catalog: .embedded)) != nil else {
          continue
        }
        received = true
      }
      try? await transport.remove(mailboxKey: mailboxKey, matchID: item.matchID)
    }
    if received { receivedToken = UUID() }
  }

  /// Un joueur tel que le destinataire l'affichera. Une photo ne voyage pas : repli sur l'emoji
  /// dérivé du pseudo, comme pour toute nouvelle fiche.
  private static func packaged(_ participant: ParticipantRecord) -> SharedMatchPackage.Participant {
    let isPhoto = participant.avatarKindSnapshot == "photo"
    let seed = LiveShareCoordinator.generatedSeed(for: participant.nicknameSnapshot)
    return SharedMatchPackage.Participant(
      participantID: participant.id,
      sharedProfileID: participant.player?.sharedProfileID,
      nickname: participant.nicknameSnapshot,
      avatarKind: isPhoto ? seed.avatarKind : participant.avatarKindSnapshot,
      avatarValue: isPhoto ? seed.avatarValue : participant.avatarValueSnapshot,
      paletteID: participant.paletteIDSnapshot)
  }
}

File: android/app/src/main/kotlin/com/quimene/app/profilesharing/SharedProfileSyncCoordinator.kt
package com.quimene.app.profilesharing

import com.quimene.app.livesync.LiveShareCoordinator
import com.quimene.domain.model.SharedMatchPackage
import com.quimene.domain.rules.GameCatalog
import com.quimene.store.MatchRepository
import com.quimene.store.ParticipantEntity
import com.quimene.store.PlayerRepository
import com.quimene.sync.MailboxCrypto
import com.quimene.sync.MailboxItem
import com.quimene.sync.MatchMailboxTransport
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.sync.Mutex
import java.util.UUID
import java.util.concurrent.atomic.AtomicBoolean

/**
 * Doc 16, phase E — miroir de `SharedProfileSyncCoordinator.swift` : historique partagé. Chaque
 * partie terminée avec un ami lié lui est déposée, **complète** et chiffrée, dans sa boîte aux
 * lettres ([MatchMailboxTransport]) ; les parties que d'autres m'ont déposées sont enregistrées ici
 * comme si je les avais jouées. Remplace les résumés du doc 14. Déclenché au lancement et au retour
 * au premier plan (`com.quimene.app.QuiMeneApplication`), dès qu'une partie se termine (écran de
 * résultats) et à l'ouverture de l'Historique (avec « tirer pour actualiser »).
 */
class SharedProfileSyncCoordinator(
    private val matchRepository: MatchRepository,
    private val playerRepository: PlayerRepository,
    private val catalog: GameCatalog,
) {
    private val transport = MatchMailboxTransport()
    private val mutex = Mutex()
    private val needsAnotherPass = AtomicBoolean(false)

    /** Une demande pendant un passage en cours n'est pas perdue : un passage de plus suit. */
    suspend fun sync() {
        if (!mutex.tryLock()) {
            needsAnotherPass.set(true)
            return
        }
        try {
            do {
                needsAnotherPass.set(false)
                deposit()
                collect()
            } while (needsAnotherPass.get())
        } finally {
            mutex.unlock()
        }
    }

    /** Une partie terminée avec au moins un ami lié (pas moi) : déposée chez chacun, une fois.
     * Jamais retentée indéfiniment pour une partie qui n'en a plus (fiche déliée entre-temps). */
    private suspend fun deposit() {
        val pending = matchRepository.matchesPendingSharedProfileSync()
        if (pending.isEmpty()) return
        val playersByID = playerRepository.allPlayers().associateBy { it.id }
        val myID = playerRepository.myOwnSharedPlayer()?.sharedProfileID

        for (match in pending) {
            val participants = matchRepository.participants(match.id)
            val profileOf = { participant: ParticipantEntity ->
                participant.playerId?.let { playersByID[it]?.sharedProfileID }
            }
            val recipients = participants.mapNotNull(profileOf).toSet() - setOfNotNull(myID)
            val events = matchRepository.currentLog(match)
            if (match.isImportedSummary || recipients.isEmpty() || events.isEmpty()) {
                matchRepository.markSharedProfileSyncComplete(match)
                continue
            }
            val pkg =
                SharedMatchPackage(
                    matchID = match.id,
                    participants = participants.map { packaged(it, profileOf(it)) },
                    events = events,
                )
            val items =
                recipients.map { recipient ->
                    MailboxItem(MailboxCrypto.lookupKey(recipient), match.id, MailboxCrypto.seal(pkg, recipient))
                }
            try {
                transport.deposit(items)
                matchRepository.markSharedProfileSyncComplete(match)
            } catch (cancellation: CancellationException) {
                throw cancellation
            } catch (error: Exception) {
                // Échec silencieux : pendingSharedProfileSync reste true, nouvelle tentative au
                // prochain déclenchement.
            }
        }
    }

    /** Ma boîte : chaque partie reçue est enregistrée (sans effet si je l'ai déjà, jouée ici ou
     * suivie dans une session), puis retirée. Un dépôt illisible est retiré aussi, pour ne pas
     * être relu indéfiniment. */
    private suspend fun collect() {
        val myID = playerRepository.myOwnSharedPlayer()?.sharedProfileID ?: return
        val mailboxKey = MailboxCrypto.lookupKey(myID)
        val items =
            try {
                transport.fetch(mailboxKey)
            } catch (cancellation: CancellationException) {
                throw cancellation
            } catch (error: Exception) {
                return
            }
        for (item in items) {
            val pkg = MailboxCrypto.open(item.ciphertext, myID)
            if (pkg != null) {
                // L'Historique observe la base : la partie y apparaît d'elle-même.
                runCatching { matchRepository.importSharedMatch(pkg, catalog) }.getOrNull() ?: continue
            }
            try {
                transport.remove(mailboxKey, item.matchID)
            } catch (cancellation: CancellationException) {
                throw cancellation
            } catch (error: Exception) {
                // Best-effort : relue plus tard (idempotent), puis purgée côté serveur.
            }
        }
    }

    /** Un joueur tel que le destinataire l'affichera. Une photo ne voyage pas : repli sur l'emoji
     * dérivé du pseudo, comme pour toute nouvelle fiche. */
    private fun packaged(
        participant: ParticipantEntity,
        profileID: UUID?,
    ): SharedMatchPackage.Participant {
        val isPhoto = participant.avatarKindSnapshot == "photo"
        val seed = LiveShareCoordinator.generatedSeed(participant.nicknameSnapshot)
        return SharedMatchPackage.Participant(
            participantID = participant.id,
            sharedProfileID = profileID,
            nickname = participant.nicknameSnapshot,
            avatarKind = if (isPhoto) seed.avatarKind else participant.avatarKindSnapshot,
            avatarValue = if (isPhoto) seed.avatarValue else participant.avatarValueSnapshot,
            paletteID = participant.paletteIDSnapshot,
        )
    }
}
